src/schemes.cpp

```diff
@@ -420,23 +420,34 @@ bool AugSchemeMPL::AggregateVerify(const vector<G1Element>& pubkeys,
     return AugSchemeMPL::AggregateVerify(pubkeys, vecMessagesBytes, signature);
 }
 
-bool AugSchemeMPL::AggregateVerify(const vector<G1Element>& pubkeys,
-                                   const vector<Bytes>& messages,
+bool AugSchemeMPL::AggregateVerify(span<G1Element const> const pubkeys,
+                                   span<Bytes const> const messages,
                                    const G2Element& signature)
 {
-    size_t nPubKeys = pubkeys.size();
-    auto arg_check = VerifyAggregateSignatureArguments(nPubKeys, messages.size(), signature);
+    if (pubkeys.size() != messages.size()) {
+        return false;
+    }
+
+    size_t const nPubKeys = pubkeys.size();
+    auto const arg_check = VerifyAggregateSignatureArguments(nPubKeys, messages.size(), signature);
     if (arg_check != CONTINUE) {
         return arg_check;
     }
 
-    vector<vector<uint8_t>> augMessages(nPubKeys);
+    std::vector<uint8_t> storage;
+    for (std::size_t i = 0; i < nPubKeys; ++i) {
+        auto const pubkey = pubkeys[i].Serialize();
+        storage.insert(storage.end(), pubkey.begin(), pubkey.end());
+        storage.insert(storage.end(), messages[i].begin(), messages[i].end());
+    }
+
+    std::vector<Bytes> augMessages;
+    augMessages.reserve(nPubKeys);
+    uint8_t const* cursor = storage.data();
     for (int i = 0; i < nPubKeys; ++i) {
-        vector<uint8_t>& aug = augMessages[i];
-        vector<uint8_t>&& pubkey = pubkeys[i].Serialize();
-        aug.reserve(pubkey.size() + messages[i].size());
-        aug.insert(aug.end(), pubkey.begin(), pubkey.end());
-        aug.insert(aug.end(), messages[i].begin(), messages[i].end());
+        auto const size = G1Element::SIZE + messages[i].size();
+        augMessages.emplace_back(cursor, size);
+        cursor += size;
```

Comment:
this patch turns one allocation per public key/message pair into a single one (or at least amortised O(1)).

This function takes a sequence of public keys and a sequence of messages and, effectively, zips them together. The only reason to do this is because the underlying API requires them to be placed next to each other. (perhaps this could be changed as well, to increase flexibility in where messages and keys are placed in memory).

The existing code makes a new heap allocation on line 437, which is inside the loop. This new allocation is where the public key and the message is copied into.

The patch copies all public keys and all messages into a single vector, which means it's a single memory object. This is likely to have these advantages:

1. there are fewer heap allocations. It could probably be improved slightly by using `malloc()` and `realloc()` instead of `std::vector<>`, but that would increase complexity quite a bit.
2. The memory is less likely to be spread out in the address space, possibly leading to better read-ahead and cache hit rate.

My benchmark suggests that the time to perform a batch verify is still dominated by the actual math. I think there may be a few more percent that could be shaved off along these lines though.

Reply:
Your proposed  code has two "small" allocations each loop iteraton i.e. each time you insert into the vector.

```
auto const pubkey = pubkeys[i].Serialize();
storage.insert(storage.end(), pubkey.begin(), pubkey.end());
storage.insert(storage.end(), messages[i].begin(), messages[i].end());
```

The old code reserves it upfront, i.e one "big" allocation each loop iteration, no? 

```
vector<uint8_t>&& pubkey = pubkeys[i].Serialize();
aug.reserve(pubkey.size() + messages[i].size());
aug.insert(aug.end(), pubkey.begin(), pubkey.end());
aug.insert(aug.end(), messages[i].begin(), messages[i].end());
```

You would probably need something like d842f14be68f2af34b121c683e7cc18eb7b958fc to complete your idea, no?

But even with that applied i don't see a noticeable speedup, i honestly don't think it makes sense to say "Benchmarks are 1.4% faster" because thats obviously in tolerance.. at least here when i run them here. If you look below i did some runs in three scenarios. 


The current `main` branch
```
Batch verification
Total: 100000 runs in 201030 ms
Avg: 2.0103 ms

Batch verification
Total: 100000 runs in 206779 ms
Avg: 2.06779 ms

Batch verification
Total: 100000 runs in 203402 ms
Avg: 2.03402 ms
```

With your patch c6e2aa7
```
Batch verification
Total: 100000 runs in 202878 ms
Avg: 2.02878 ms

Batch verification
Total: 100000 runs in 203268 ms
Avg: 2.03268 ms

Batch verification
Total: 100000 runs in 221848 ms
Avg: 2.21848 ms
```

With your patch c6e2aa7 and my patch d842f14be68f2af34b121c683e7cc18eb7b958fc on top 

```
Batch verification
Total: 100000 runs in 218086 ms
Avg: 2.18086 ms

Batch verification
Total: 100000 runs in 200706 ms
Avg: 2.00706 ms

Batch verification
Total: 100000 runs in 209845 ms
Avg: 2.09845 ms
```

As you can see they more or less all the same.

Reply:
> Your proposed code has two "small" allocations each loop iteraton i.e. each time you insert into the vector.

No. vector has amortized O(1) cost of insertion. In practice it means it will grow the size of the vector proportional to the current size, not proportional to what's being inserted.

> The old code reserves it upfront, i.e one "big" allocation each loop iteration, no?

No. It reserves a vector *of vectors*. Only the top level vector is reserved, then all the other vectors are allocated indivudually.

> i honestly don't think it makes sense to say "Benchmarks are 1.4% faster"

Yeah, that's fair. I did not measure very rigorously. I think it would make sense to remove this step entirely though. It seems like that might be possible, and that it's only here because of somewhat arbitrary interface layers stacked, requiring massaging of the data.

Reply:
I backed this out and will make another attempt in a separate PR